src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/InlineCompletion/InlineCompletionEndPoint.cs

```diff
@@ -0,0 +1,240 @@
+        if (list == null)
+        {
+            _logger.LogInformation($"Did not get any inline completions from delegation.");
+            return null;
+        }
+
+        var items = new List<InlineCompletionItem>();
+        var csharpDocOptions = codeDocument.GetCSharpDocument();
+        foreach (var item in list.Items)
+        {
+            var containsSnippet = item.TextFormat == InsertTextFormat.Snippet;
+            var range = item.Range ?? new Range { Start = projectedPosition, End = projectedPosition };
+
+            if (!_documentMappingService.TryMapFromProjectedDocumentRange(codeDocument, range, out var rangeInRazorDoc))
+            {
+                _logger.LogWarning($"Could not remap projected range {range} to razor document");
+                continue;
+            }
+
+            using var formattingContext = FormattingContext.Create(request.TextDocument.Uri, document, codeDocument, s_defaultFormattingOptions, _adhocWorkspaceFactory, isFormatOnType: true, automaticallyAddUsings: false);
+            if (!TryGetSnippetWithAdjustedIndentation(formattingContext, item.Text, hostDocumentIndex, out var newSnippetText))
+            {
+                continue;
+            }
+
+            var remappedItem = new InlineCompletionItem
+            {
+                Command = item.Command,
+                Range = rangeInRazorDoc,
+                Text = newSnippetText.ToString(),
+                TextFormat = item.TextFormat,
+            };
+            items.Add(remappedItem);
+        }
+
+        _logger.LogInformation($"Returning items.");
+        return new InlineCompletionList
+        {
+            Items = items.ToArray()
+        };
+    }
+
+    private static bool TryGetSnippetWithAdjustedIndentation(FormattingContext formattingContext, string snippetText, int hostDocumentIndex, [NotNullWhen(true)] out string? newSnippetText)
+    {
+        newSnippetText = null;
+        if (!formattingContext.TryGetFormattingSpan(hostDocumentIndex, out var formattingSpan))
+        {
+            return false;
+        }
+
+        // Take the amount of indentation razor and html are adding, then remove the amount of C# indentation that is 'hidden'.
+        // This should give us the desired base indentation that must be applied to each line.
+        var razorAndHtmlContributionsToIndentation = formattingSpan.RazorIndentationLevel + formattingSpan.HtmlIndentationLevel;
+        var amountToAddToCSharpIndentation = razorAndHtmlContributionsToIndentation - formattingSpan.MinCSharpIndentLevel;
+
+        var snippetSourceText = SourceText.From(snippetText);
+        List<TextChange> indentationChanges = new();
+        // Adjust each line, skipping the first since it must start at the snippet keyword.
+        foreach (var line in snippetSourceText.Lines.Skip(1))
+        {
+            var text = snippetSourceText.GetSubText(line.Span).ToString();
```

Comment:
Instead of ToStringing the subtext would we be able to do `if (subText.Length == 0)` below?